### PR TITLE
[MAINT] use tmp_path in tests for first lvl GLM

### PR DESCRIPTION
**Pull request by Remi-Gau**
- relates to https://github.com/nilearn/nilearn/issues/3726
- extracted from #3891
- ~~blocked by #3892~~

### CODE REVIEW

File: nilearn/glm/tests/test_first_level.py [hunk 1]
```diff
@@ -89,137 +88,134 @@ def test_high_level_glm_one_session():
...
+    assert_almost_equal(
+        get_data(fixed_fx_contrast),
+        get_data(fixed_fx_dic['effect_size']))
+    assert_almost_equal(
+        get_data(fixed_fx_variance),
+        get_data(fixed_fx_dic['effect_variance']))
+    assert_almost_equal(
+        get_data(fixed_fx_stat), get_data(fixed_fx_dic['stat']))
+
+    # test without mask variable
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:133]: parametrize?

File: nilearn/glm/tests/test_first_level.py [hunk 2]
```diff
@@ -89,137 +88,134 @@ def test_high_level_glm_one_session():
...
+    assert_array_equal(get_data(all_images['z_score']), get_data(z_image))
+    assert_array_equal(get_data(all_images['p_value']), get_data(p_value))
+    assert_array_equal(get_data(all_images['stat']), get_data(stat_image))
+    assert_array_equal(get_data(all_images['effect_size']),
+                       get_data(effect_image))
+    assert_array_equal(get_data(all_images['effect_variance']),
+                       get_data(variance_image))
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del (all_images,
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:192]: still matters with tmp_path?

File: nilearn/glm/tests/test_first_level.py [hunk 3]
```diff
@@ -89,137 +88,134 @@ def test_high_level_glm_one_session():
...
+    shapes, rk = ((7, 8, 7, 15), (7, 8, 7, 16)), 3
+    mask, fmri_data, design_matrices =\
+        write_fake_fmri_data_and_design(shapes, rk, file_path=tmp_path)
+    multi_session_model = FirstLevelModel(mask_img=None).fit(
+        fmri_data, design_matrices=design_matrices)
+    n_voxels = get_data(multi_session_model.masker_.mask_img_).sum()
+    z_image = multi_session_model.compute_contrast(np.eye(rk)[1])
+    assert np.sum(get_data(z_image) != 0) == n_voxels
+    assert get_data(z_image).std() < 3.
+    # with mask
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:165]: extract into separate test

File: nilearn/glm/tests/test_first_level.py [hunk 4]
```diff
@@ -89,137 +88,134 @@ def test_high_level_glm_one_session():
...
+    mask_file, fmri_files, design_files =\
+        write_fake_fmri_data_and_design(shapes, rk, file_path=tmp_path)
+    multi_session_model = FirstLevelModel(mask_img=None).fit(
+        fmri_files, design_matrices=design_files)
+    z_image = multi_session_model.compute_contrast(np.eye(rk)[1])
+    assert_array_equal(z_image.affine, load(mask_file).affine)
+    assert get_data(z_image).std() < 3.
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del z_image, fmri_files, multi_session_model
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:218]: still necessary?

File: nilearn/glm/tests/test_first_level.py [hunk 5]
```diff
@@ -432,149 +428,152 @@ def test_scaling():
...
+    conf = pd.DataFrame([0, 0])
+
+    events = basic_paradigm()
+    for fi in func_img, FUNCFILE:
+        for d in des, des_fname:
+            FirstLevelModel().fit(fi, design_matrices=d)
+            FirstLevelModel(mask_img=None).fit([fi], design_matrices=d)
+            FirstLevelModel(mask_img=mask).fit(fi, design_matrices=[d])
+            FirstLevelModel(mask_img=mask).fit([fi], design_matrices=[d])
+            with pytest.warns(UserWarning, match="If design matrices "
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:453]: extract negative tests

File: nilearn/glm/tests/test_first_level.py [hunk 6]
```diff
@@ -432,149 +428,152 @@ def test_scaling():
...
+    design = make_first_level_design_matrix(frame_times, events,
+                                            drift_model='polynomial',
+                                            drift_order=3)
+    frame2, X2, names2 = check_design_matrix(design)
+    assert_array_equal(frame1, frame2)
+    assert_array_equal(X1, X2)
+    assert_array_equal(names1, names2)
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del FUNCFILE, mask, model, func_img
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:534]: still needed ?

File: nilearn/glm/tests/test_first_level.py [hunk 7]
```diff
@@ -432,149 +428,152 @@ def test_scaling():
...
+    events = basic_paradigm()
+    # Ordinary Least Squares case
+    model = FirstLevelModel(t_r, slice_time_ref, mask_img=mask,
+                            drift_model='polynomial', drift_order=3,
+                            minimize_memory=False)
+    model = model.fit(func_img, events)
+
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del mask, FUNCFILE, func_img, model
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:555]: still needed?

File: nilearn/glm/tests/test_first_level.py [hunk 8]
```diff
@@ -432,149 +428,152 @@ def test_scaling():
...
+    events = basic_paradigm()
+    # Ordinary Least Squares case
+    model = FirstLevelModel(t_r, slice_time_ref, mask_img=mask,
+                            drift_model='polynomial', drift_order=3,
+                            memory='nilearn_cache', memory_level=1,
+                            minimize_memory=False)
+    model.fit(func_img, events)
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del mask, func_img, FUNCFILE, model
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:576]: still nedded ?

File: nilearn/glm/tests/test_first_level.py [hunk 9]
```diff
@@ -773,66 +772,65 @@ def test_first_level_from_bids_get_StartTime_from_derivatives(tmp_path):
...
+    t_r = 10.0
+    slice_time_ref = 0.
+    events = basic_paradigm()
+    # Ordinary Least Squares case
+    model = FirstLevelModel(t_r, slice_time_ref, mask_img=mask,
+                            drift_model='polynomial', drift_order=3,
+                            minimize_memory=False)
+    c1, c2, cnull = np.eye(7)[0], np.eye(7)[1], np.zeros(7)
+    # asking for contrast before model fit gives error
+    with pytest.raises(ValueError):
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:791]: extract negative tests

File: nilearn/glm/tests/test_first_level.py [hunk 10]
```diff
@@ -773,66 +772,65 @@ def test_first_level_from_bids_get_StartTime_from_derivatives(tmp_path):
...
+        model.compute_contrast([])
+    with pytest.raises(ValueError):
+        model.compute_contrast([c1, []])
+    with pytest.raises(ValueError):
+        model.compute_contrast(c1, '', '')
+    with pytest.raises(ValueError):
+        model.compute_contrast(c1, '', [])
+    # Delete objects attached to files to avoid WindowsError when deleting
+    # temporary directory (in Windows)
+    del func_img, FUNCFILE, model
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:833]: still needed?

**Comment by bthirion** [comment]
LMK when it deserves a review.

**Comment by Remi-Gau** [comment]
> LMK when it deserves a review.

not before #3892 is in because this PR depends on it

**Review by ymzayek — APPROVED** [review]
Thx @Remi-Gau

**Comment by Remi-Gau** [comment]
Thanks @ymzayek

**Pull request merged**